c_src/sasl_auth.c

```diff
@@ -166,12 +302,17 @@ static ERL_NIF_TERM sasl_list_mech(ErlNifEnv* env, int UNUSED(argc), const ERL_N
 
     if ((!enif_get_resource(env, argv[0], sasl_resource, (void**)&state))) {
         return enif_make_badarg(env);
+    } else if (!sasl_auth_process_check(env, state)) {
+        return enif_raise_exception(env, ATOM_NOT_CONTROLLING_PROCESS);
     }
 
+    enif_mutex_lock(state->controller_lock);
```

Comment:
The additional usage of the controller lock may not be needed anymore, simply per the check of the controlling process in conjunction with not exposing a method for changing the controlling process. However, I had concern over the use of global state in cyrus-sasl.